lib/mobius/core/command.ex

```diff
@@ -34,49 +40,44 @@ defmodule Mobius.Core.Command do
   @spec arg_count(t()) :: non_neg_integer()
   def arg_count(%__MODULE__{} = command), do: length(command.args)
 
-  @spec execute_command([t()], String.t(), Message.t()) :: handle_message_result()
+  @spec execute_command(processed(), String.t(), Message.t()) :: handle_message_result()
   def execute_command(commands, prefix, %Message{content: content} = message) do
-    arg_values = split_arguments(content)
-
-    Enum.reduce_while(commands, :not_a_command, fn %__MODULE__{} = command, acc ->
-      new =
-        cond do
-          not String.starts_with?(content, prefix <> command.name) -> :not_a_command
-          arg_count(command) != length(arg_values) -> {:too_few_args, command, length(arg_values)}
-          true -> try_command(command, arg_values, message)
-        end
-
-      decide_by_priority(acc, new)
-    end)
+    with {:ok, content} <- match_prefix(prefix, content),
+         {:ok, name, arg_values} <- split_arguments(content),
+         {:ok, groups} <- get_command(commands, name),
+         {:ok, clauses} <- get_clauses(groups, length(arg_values)),
+         {:ok, command, values} <- find_clause(clauses, arg_values) do
+      {:ok, apply(command.handler, [message | values])}
+    end
   end
 
-  defp decide_by_priority(old, new) do
-    # Overwrite the accumulator depending on a priority list:
-    # :ok (which halts the reduce) > :invalid_args > :too_few_args > :not_a_command
-    # All the :cont are to let other clauses have their chance
-    # However if one of the clauses has the right amount of args, but wrong types, that's the
-    #   error we want to return ultimately if no other clause matches those arguments
-    # Same logic applies for :too_few_args, if one clause is the right command with the wrong
-    #   amount of args, but no other clause matches the command, that's the error we want
-    # Finally, if none of the commands matches, we return :not_a_command
-    case {old, new} do
-      {{:ok, _} = value, _} -> {:halt, value}
-      {_, {:ok, _} = value} -> {:halt, value}
-      {{:invalid_args, _} = value, _} -> {:cont, value}
-      {_, {:invalid_args, _} = value} -> {:cont, value}
-      {{:too_few_args, _, _} = value, _} -> {:cont, value}
-      {_, {:too_few_args, _, _} = value} -> {:cont, value}
-      {:not_a_command, _} -> {:cont, :not_a_command}
-      {_, :not_a_command} -> {:cont, :not_a_command}
+  @spec preprocess_commands([t()]) :: processed()
+  def preprocess_commands(commands) do
+    commands
+    |> Enum.reverse()
+    |> Enum.group_by(fn %__MODULE__{name: name} -> name end)
+    |> Enum.map(fn {name, commands} -> {name, Enum.group_by(commands, &arg_count/1)} end)
+    |> Map.new()
```

Comment:
```suggestion
    |> Map.new(fn {name, commands} -> {name, Enum.group_by(commands, &arg_count/1)} end)
```
The line feels a little dense with the group_by call inside the transform but I can't think of a cleaner way to do this 🤔

Reply:
I always forget about `new` taking a function 🤦

And yeah I agree it's a bit dense, but I tried extracting parts of that line to a function and it just felt like I was making spaghetti so I moved it back instead